## Group SQL Practice

_Author: Matt Brems & Katie Sylvia_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select count(*) from stores
"""
# with engine.connect() as conn:
#    for row in conn.execute(sql):
#        print(f'The number of rows in the stores table are: {row[0]}')

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select name from stores order by name limit 5
"""

pd.read_sql_query(sql, engine)[3:4]

name
3  3061 Cub Foods / Sioux City

**Prompt 2 Answer**:

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [16]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select vendor, btl_price, bottle_qty 
from sales 
where category_name = 'IMPORTED VODKA'
order by vendor 
limit 3
"""

pd.read_sql_query(sql, engine)

vendor btl_price  bottle_qty
0  Adamba Imports Int'l Inc.     $8.99          12
1      A Hardy / U.S.A. Ltd.    $28.50           6
2      A Hardy / U.S.A. Ltd.    $28.50           3

**Prompt 3 Answer**: 


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [20]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select sum(total)
from sales
where date >= '2015-02-01'
 and  date < '2015-03-01'
"""
pd.read_sql_query(sql, engine).map('{:,.0f}'.format)

sum
0  21,295,351

**Prompt 4 Answer**:

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [70]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select vendor, 
       count(store) as tranx_count,
       sum(bottle_qty) as tot_bottles,
       sum(total) as revenue
from sales 
where category_name = 'IMPORTED VODKA'
group by vendor
order by tranx_count desc
limit 3
"""

pd.read_sql_query(sql, engine)

vendor  tranx_count  tot_bottles     revenue
0  Pernod Ricard USA/Austin Nichols        42338       403450  7880365.88
1               Bacardi U.S.A. Inc.        26441       236047  5771242.07
2                   Diageo Americas        15725       152038  3346425.42

**Prompt 5 Answer**: 

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [25]:
sql = """
select btl_price - state_btl_cost as markup 
from sales 
limit 3
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [73]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select st.name as store_name,
       (avg((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost)*100) as avg_pct_markup
from sales sa, stores st
where st.store = sa.store
group by store_name
order by avg_pct_markup desc
limit 10
"""

pd.read_sql_query(sql, engine)

store_name  avg_pct_markup
0        Wal-Mart 0841 / Tipton       52.547124
1    Wal-Mart 1546 / Iowa Falls       52.149880
2    Wal-Mart 1683 / Shenandoah       51.890342
3      Wal-Mart 3394 / Atlantic       51.509378
4     Wal-Mart 2935 / Knoxville       51.459084
5  Wal-Mart 1528 / Cedar Rapids       51.443853
6       Wal-Mart 1005 / Waverly       51.412550
7     Wal-Mart 1509 / Maquoketa       51.409767
8    Wal-Mart 1526 / Storm Lake       51.374724
9    Wal-Mart 1723 / Des Moines       51.328566

**Prompt 7 Answer**: 

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [30]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select st.name as store_name,
       st.store_address, 
       st.store as store_id,
       avg((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) as avg_pct_markup,
       min((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) as min_pct_markup,
       max((sa.btl_price - sa.state_btl_cost) / sa.state_btl_cost) as max_pct_markup
from sales sa, stores st
where st.store = sa.store
group by st.name, st.store_address, st.store
order by avg_pct_markup desc
limit 5
"""

pd.read_sql_query(sql, engine)

store_name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1546 / Iowa Falls   
2  Wal-Mart 1683 / Shenandoah   
3    Wal-Mart 3394 / Atlantic   
4   Wal-Mart 2935 / Knoxville   

                                       store_address  store_id  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...      4013   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...      4024   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...      4266   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....      3833   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...      3660   

   avg_pct_markup  min_pct_markup  max_pct_markup  
0        0.525471        0.498571        6.708763  
1        0.521499        0.498462        6.708763  
2        0.518903        0.498462        6.708763  
3        0.515094        0.498462        6.708763  
4        0.514591        0.498462        6.708763

**Prompt 8 Answer**:

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [48]:
sql = """
select c.county,
      (sum(s.bottle_qty))/c.population bottles_per_capita,
      round(avg(p.bottle_size),1) as avg_bottle_size
from counties c, sales s, products p
where c.county = s.county
 and  p.item_no = s.item
group by c.county
order by bottles_per_capita desc
limit 9
"""

pd.read_sql_query(sql, engine)

county  bottles_per_capita  avg_bottle_size
0      Dickinson                  22            974.1
1           Polk                  15            892.6
2     Black Hawk                  15            862.9
3    Cerro Gordo                  14            986.4
4          Scott                  13            871.6
5        Johnson                  13            895.4
6           Linn                  12            886.3
7  Pottawattamie                  11            893.7
8        Kossuth                  11           1041.0

##### Answer to question:
Kossuth county has the largest average bottle size of 1041 ml.

**Prompt 9 Answer**:

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [66]:
# Use a SQL query to find the answer to the above prompt.
# ha. Trick question.  None.
sql = """
select count(vendor)
from sales
where total::numeric <> (btl_price * bottle_qty)::numeric

"""

pd.read_sql_query(sql, engine)

count
0      0

**Prompt 10 Answer**: